iopipe/iopipe.py

```diff
@@ -126,3 +128,29 @@ def timeout_handler(self, report, signum, frame):
         """
         logger.debug('Function is about to timeout, sending report')
         report.send()
+
+    def load_plugins(self, plugins):
+        """
+        Loads plugins that match the `Plugin` interface and are instantiated.
+
+        :param plugins: A list of plugin instances.
+        """
+        def instantiate(plugin):
+            return plugin() if inspect.isclass(plugin) else plugin
+
+        return [instantiate(p) for p in plugins if is_plugin(p)]
+
+    def run_hooks(self, name, event=None, context=None):
+        """
+        Runs plugin hooks for each registered plugin.
+        """
+        hooks = {
+            'setup': lambda p: p.setup(self),
```

Comment:
Why isn't there a pre: and post: setup?

Reply:
@pselle Because setting plugins is coupled with the rest of agent config. It didn't make sense to have a pre-config hook (`pre:setup`). All the python agent really does at "setup" is sort out config, and I'm unclear on what a plugin would even need to do pre-config considering plugins themselves are configured that way. There is also error handling in the node.js agent regarding `pre:setup` which seems more complicated than just having a single hook at initialization.

https://github.com/iopipe/iopipe/blob/master/src/index.js#L165

Reply:
pre:setup would be involved if someone wanted to modify the config object itself, before it's configured, for example. In any case, this is a larger architecture decision to introduce inconsistency across agents ... we should have pre: and post: for consistency.